experimental/torch_xla2/examples/train_llama_torchtitan/train_llama.py

```diff
@@ -0,0 +1,283 @@
+        num_tokens = logits.shape[-1]
+        logits = logits.reshape(-1, num_tokens)
+        y = y.reshape(-1)
+        return torch.nn.functional.cross_entropy(
+            logits, y)
+    
+
+class Trainer:
+
+    def __init__(self):
+        self.mesh = jax.sharding.Mesh(
+            mesh_utils.create_device_mesh(num_partitions),
+            axis_names=global_axis,
+        )
+        self.x_sharding = jax.sharding.NamedSharding(self.mesh, P(global_axis))
+        self.y_sharding = jax.sharding.NamedSharding(self.mesh, P(*global_axis))
+        self.replicated = jax.sharding.NamedSharding(self.mesh, P())
+
+    def torch_opt_to_jax_opt(self, torch_opt):
+        # TODO: Can convert optimizer instead of using a jax one
+        return optax.adamw(0.01)
+
+    def _shard_fsdp_style(self, state_dict, sharding=None):
+        if sharding is None:
+            sharding = self.x_sharding
+        def move_one_tensor(x):
+            jval = torch_xla2.tensor.t2j(x)
+            return sharded_device_put(jval, sharding)
+
+        if isinstance(state_dict, torch.Tensor):
+            return move_one_tensor(state_dict)
+        res = {}
+        for k, v in sorted(state_dict.items()):
+            res[k] = move_one_tensor(v)
+        return res
+
+    def fit(self, lightning_mod, data_loader):
+        xla_env = torch_xla2.default_env()
+        jax.config.update('jax_enable_x64', False)
+        xla_env._mesh = self.mesh
+        xla_env.use_flash_attention = True
+
+        jittable_mod = JittableModule(lightning_mod)
+        jax_params = self._shard_fsdp_style(jittable_mod.params) 
+        jax_buffers = self._shard_fsdp_style(jittable_mod.buffers)
+
+        @jax.checkpoint
+        def lightning_mod_loss(
+            weights: jax.Array, buffers: jax.Array, data: jax.Array, batch_id):
+            """returns loss"""
+            with jax.named_scope("Computing_loss"):
+                weights, buffers, data = torch_view((weights, buffers, data))
+                # NOTE: these is needed because the original model
+                # did not register those as persistent buffer
+                with xla_env:
+                    loss = jittable_mod.functional_call(
+                        'training_step',
+                        weights, buffers, data, batch_id)
+                return jax_view(loss)
+
+        jax_optimizer = self.torch_opt_to_jax_opt(None)
+
+        opt_state = jax_optimizer.init(jax_params)
+        grad_fn = jax.value_and_grad(lightning_mod_loss)
+
+        opt_state_sharding = jax.tree_util.tree_map(lambda p : p.sharding, opt_state)
+
+        print('Begining training')
+
+        # NOTE: explicitly set sharding so the sharding of opt_state wont change
+        # if it changes, it would trigger recompile
+        @functools.partial(
+            jax.jit, 
+            donate_argnums=(0, 2),
+            #in_shardings=(self.x_sharding, self.x_sharding, opt_state_sharding, self.x_sharding, self.replicated),
```

Comment:
nit: is this explicit sharding needed? the comment on line 167 says it's required

Reply:
removed.